### PR TITLE
WIP - Add Button Dropdown Component

### PR DESCRIPTION
**Pull request by kwood15**
Please have a look over what I have done so far, I have more to do on this.

### CODE REVIEW

**Comment by kwood15** [comment]
Have finished this now so if you can have a look over what I have done. I have not put any grid/layout in as per our other components. Have included  below Ignore the overlap of spacing as there is no grid/gutters. Will need to write some unit tests for these also once reviewed. Cheers

screen shot 2017-08-30 at 17 32 22

screen shot 2017-08-30 at 17 32 44

screen shot 2017-08-30 at 17 33 02

File: demo/src/App.jsx [hunk 1]
```diff
@@ -153,8 +154,54 @@ class App extends Component {
               </Flyout.Content>
             </Flyout.Wrapper>
             <Button type="submit" size="md" theme="beta">
-            Button test
-          </Button>
+              Button test
+            </Button>
+            <h3>Button Dropdowns</h3>
+            <ButtonDropdown.Wrapper togglePosition="left">
```
Comment (rbrtsmith) [inline demo/src/App.jsx]: API for button dropdown looks good.  This is pretty much what we want to present to consumers.

File: package.json [hunk 2]
```diff
@@ -30,6 +30,7 @@
   ],
   "dependencies": {
     "babel-plugin-transform-object-rest-spread": "^6.23.0",
+    "classnames": "^2.2.5",
```
Comment (rbrtsmith) [inline package.json]: We already have a local version of classNames in utils.  We're not looking to add any dependencies here so can you remove this dependency.
But that's also alerted me to `babel-plugin-transform-object-rest-spread`
Could you move that to `devDependencies` please?

Comment (kwood15) [reply package.json]: Will do

File: src/components/ButtonDropdown/Toggle.jsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { createElement as E } from 'react'
+import T from 'prop-types'
+import { classNames } from '../../utils/'
+
+const initial = 'c-btn-dropdown__toggle'
```
Comment (rbrtsmith) [inline src/components/ButtonDropdown/Toggle.jsx]: don't really need this in a variable, it's only used once so just put the string as a literal in `classNames`

File: src/components/ButtonDropdown/Wrapper.jsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      }
+      return child
+    })
+    return (
+      E(
+        node || 'div',
+        {
+          className: enhancedClassName,
+          // eslint-disable-next-line no-shadow
+          ref: (node) => { this.wrapperRef = node }
```
Comment (rbrtsmith) [inline src/components/ButtonDropdown/Wrapper.jsx]: replace `node` with `n` then we don't need the `eslint disable`.  It's important that we don't shadow variables.

Comment (kwood15) [reply src/components/ButtonDropdown/Wrapper.jsx]: :)

File: src/components/ButtonDropdown/Wrapper.jsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      )
+    )
+  }
+}
+
+ButtonDropdownWrapper.propTypes = {
+  node: T.string,
+  className: T.string,
+  children: T.node.isRequired,
+  togglePosition: T.oneOf(['left', 'right'])
```
Comment (rbrtsmith) [inline src/components/ButtonDropdown/Wrapper.jsx]: good.

File: package.json (hunk 2 above)
Comment (kwood15) [reply package.json]: done

File: src/components/ButtonDropdown/Wrapper.jsx (hunk 4 above)
Comment (kwood15) [reply src/components/ButtonDropdown/Wrapper.jsx]: done

File: src/components/ButtonDropdown/Toggle.jsx (hunk 3 above)
Comment (kwood15) [reply src/components/ButtonDropdown/Toggle.jsx]: done

File: src/components/ButtonDropdown/Content.jsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import T from 'prop-types'
+import { classNames } from '../../utils/'
+
+const initial = 'c-btn-dropdown__content'
+
+const ButtonDropdownContent = ({ node, className, children, ...rest }) =>
+  E(
+    node || 'div',
+    {
+      className: classNames(initial, className),
```
Comment (rbrtsmith) [inline src/components/ButtonDropdown/Content.jsx]: Initial still in a variable.

**Review by rbrtsmith — CHANGES_REQUESTED** [review]
Requires unit tests now.

File: src/components/ButtonDropdown/Content.jsx (hunk 6 above)
Comment (kwood15) [reply src/components/ButtonDropdown/Content.jsx]: done

**Comment by kwood15** [comment]
Last fix done now

**Comment by kwood15** [comment]
Have added some unit tests for these now - I have had to skip 3 tests as these aren't working so could do with your help as to why so if you get a min. Thanks

**Comment by kwood15** [comment]
Please have look over this - have added additional tests to the wrapper

**Review by rbrtsmith — APPROVED** [review]

**Pull request merged**